integration/integration_test.go

```diff
@@ -545,27 +545,142 @@ func TestBuildFromDevcontainerWithFeatures(t *testing.T) {
 	require.Equal(t, "hello from test 3!", strings.TrimSpace(test3Output))
 }
 
-func TestBuildFromDockerfile(t *testing.T) {
-	// Ensures that a Git repository with a Dockerfile is cloned and built.
-	srv := gittest.CreateGitServer(t, gittest.Options{
-		Files: map[string]string{
-			"Dockerfile": "FROM " + testImageAlpine,
+func TestBuildFromDockerfileAndConfig(t *testing.T) {
+	t.Parallel()
+
+	type configFile struct {
+		name string
+		data string
+	}
+	type testCase struct {
+		name         string
+		env          []string
+		configFile   configFile
+		configBase64 string
+		validate     func(t *testing.T, tc testCase, ctrID, logs string)
+	}
+
+	validateDockerConfig := func(t *testing.T, tc testCase, ctrID, logs string) {
+		t.Helper()
+
+		// Ensure that the config matches the expected value, base64 is
+		// always prioritized over a file.
+		got := execContainer(t, ctrID, "cat /docker_config_json")
+		got = strings.TrimSpace(got)
+		want := tc.configBase64
+		if want == "" {
+			want = tc.configFile.data
+		}
+		if want != "" {
+			require.Contains(t, logs, "Set DOCKER_CONFIG to /.envbuilder/.docker")
+			require.Equal(t, want, got)
+		}
+
+		// Ensure that a warning message is printed if config secrets
```

Comment:
Why do we only warn on this? Users may have a reason to use a DOCKER_CONFIG at build time and they may have a reason to use one at run time. But those are separate use cases as far as I can tell. Should we allow the line between the two to blur?

Would it be more secure and user friendly to treat build time DOCKER_CONFIG like a build secret; ensuring that it's always cleared?

Runtime secrets could then be provided using a runtime mechanism? Right users are mounting it in, but to me this would be treated just like any other runtime secret.

Reply:
> Why do we only warn on this? Users may have a reason to use a DOCKER_CONFIG at build time and they may have a reason to use one at run time. But those are separate use cases as far as I can tell. Should we allow the line between the two to blur?

That's a valid question. We can't determine the user's intent here. I'd say it's more likely that a given `DOCKER_CONFIG` is intended for build time since there's no guarantee Docker will be available in the resulting devcontainer.

Adding to this, if the devcontainer should have a `DOCKER_CONFIG` set at runtime, that should be done via `containerEnv` in `devcontainer.json`, not on the envbuilder container.

> Would it be more secure and user friendly to treat build time DOCKER_CONFIG like a build secret; ensuring that it's always cleared?

Perhaps, but this ties into why we "only warn". We can't guarantee that we'll be able to wipe the secrets if the user has mounted them in, possibly as read-only, and the container doesn't have unmount permissions (or, will remounting be possible if there's sudo?). That's why we only warn and don't attempt to clear. If we attempted something else, we may risk introducing behavior that varies depending on the environment it's run in. It's better if the user knows what to expect.

Also, since we restore the original `DOCKER_CONFIG`, we're allowing that config location to be used at runtime, enabling use-cases where you have a separate build and runtime config. You set the build config via base64 env (takes precedence), and after the build, the location will be used at runtime. As a side-effect, the same config can be used for _both build and runtime_, which might not be ideal, but what's the alternative?

> Runtime secrets could then be provided using a runtime mechanism? Right users are mounting it in, but to me this would be treated just like any other runtime secret.

Not sure what you're referring to by runtime secret and runtime mechanism? Could you expand on this?